warehouse/logfield/logfield.go

```diff
@@ -23,7 +23,82 @@ const (
 	LoadFileType               = "loadFileType"
 	ErrorMapping               = "errorMapping"
 	DestinationCredsValid      = "destinationCredsValid"
-	Query                      = "query"
 	QueryExecutionTime         = "queryExecutionTime"
 	StagingTableName           = "stagingTableName"
 )
+
+// TODO finish converting constants to functions to experiment
+
+type Field func(v interface{}) KeyValue
+
+type KeyValue struct {
+	k string
+	v interface{}
+}
+
+var (
+	UploadID      Field = func(v interface{}) KeyValue { return KeyValue{"uploadID", v} }
+	SourceID      Field = func(v interface{}) KeyValue { return KeyValue{"sourceID", v} }
+	DestinationID Field = func(v interface{}) KeyValue { return KeyValue{"destinationID", v} }
+	TableName     Field = func(v interface{}) KeyValue { return KeyValue{"tableName", v} }
+	Location      Field = func(v interface{}) KeyValue { return KeyValue{"location", v} }
+	Query         Field = func(v interface{}) KeyValue { return KeyValue{"query", v} }
+	Error         Field = func(v interface{}) KeyValue {
+		if err, ok := v.(error); ok {
+			return KeyValue{"error", err.Error()}
+		} else {
+			return KeyValue{"error", v}
+		}
+	}
+
+	// Generic should be used only when there is no other option and it doesn't make sense to create a new Field
+	Generic = func(k string) Field {
+		return func(v interface{}) KeyValue {
+			return KeyValue{k, v}
+		}
+	}
+)
+
+func NewLogger(component string, l logger.Logger) *Logger {
+	return &Logger{component: component, logger: l}
+}
+
+type Logger struct {
```

Comment:
[optional] Given that the `Logger` is a struct here we could always pass it by value and then use a `sync.Once` to make sure that the internal logger is always either not `nil` or `logger.NOP`. This would help with the tests as well.

Reply:
Should we enforce this at the `logger.Logger`?

Reply:
> How are we planning to use this? Are we going to replace logger.Logger with this one everywhere?

That would be the idea yeah. Of course it'll be done incrementally and with time.

> Should we enforce this at the `logger.Logger`?

The `logger.Logger` is generic and it belongs to the `kit`. I was thinking of a logger wrapper that is domain/team specific. Such logger could even live in an `internal` folder and used only in its specific domain. Also, the abstraction layer is very small, it's mostly about providing the fields for consistency so I don't see a problem with having a package like this in each domain.